node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java

```diff
@@ -194,27 +194,34 @@ private boolean offeredNodeHasParentHostnameAlreadyAccepted(NodeCandidate candid
         return false;
     }
 
+    /**
+     * Returns whether allocating the candidate on this host would violate exclusivity constraints.
+     * Note that while we currently require that exclusive allocations uses the entire host,
+     * this method also handles the case where smaller exclusive nodes are allocated on it.
+     */
     private boolean violatesExclusivity(NodeCandidate candidate) {
-        if (candidate.parentHostname().isEmpty()) return false;
-
-        // In nodes which does not allow host sharing, exclusivity is violated if...
-        if ( ! nodeRepository.zone().cloud().allowHostSharing()) {
-            // TODO: Write this in a way that is simple to read
-            // If either the parent is dedicated to a cluster type different from this cluster
-            return  ! candidate.parent.flatMap(Node::exclusiveToClusterType).map(cluster.type()::equals).orElse(true) ||
-                    // or this cluster is requiring exclusivity, but the host is exclusive to a different owner
-                    (requestedNodes.isExclusive() && !candidate.parent.flatMap(Node::exclusiveToApplicationId).map(application::equals).orElse(false));
+        if (candidate.parent.isEmpty()) return false;
+
+        if (nodeRepository.exclusiveAllocation(cluster)) {
+            // Node must allocate the host entirely and not violate application or cluster type constraints
+            var parent = candidate.parent.get();
+            if (!candidate.resources().isUnspecified() &&
+                ! nodeRepository.resourcesCalculator().advertisedResourcesOf(parent.flavor()).compatibleWith(candidate.resources())) return true;
+            if (parent.exclusiveToApplicationId().isPresent() && !parent.exclusiveToApplicationId().get().equals(application)) return true;
```

Comment:
If we require exclusivity, our host should have exclusiveToApplicationId set, so this should be 
```suggestion
            if (parent.exclusiveToApplicationId().isEmpty() || !parent.exclusiveToApplicationId().get().equals(application)) return true;
```

Reply:
That we need this PR at all is proof that we are offered nodes on hosts which are not exclusive. Perhaps that should be fixed elsewhere.

With the "fills completely" policy there is no need for exclusiveToApplicationId, though, so perhaps we should remove that property (although not in _this_ PR). In any case, I think this method should enforce the actual constraints regardless of what might obtain at some point in time given the rest of the code (and live pre-existing state). And I think it does with these three constraints?

Reply:
Is there more to discuss here?